File: renderer/shadow.py
import cv2
import numpy as np


def make_shadow(mask: np.ndarray, cfg: dict) -> np.ndarray:
    """
    Generate a soft drop shadow from the painting mask.

    Returns an RGBA image (same size as mask) representing the shadow.
    """
    blur = int(cfg.get("blur", 18))
    opacity = float(cfg.get("opacity", 0.18))
    offset_x = int(cfg.get("offsetX", 8))
    offset_y = int(cfg.get("offsetY", 10))

    h, w = mask.shape

    # Shift mask by offset
    shifted = np.zeros_like(mask)
    y_start = max(0, offset_y)
    y_end = min(h, h + offset_y)
    x_start = max(0, offset_x)
    x_end = min(w, w + offset_x)

    src_y_start = max(0, -offset_y)
    src_y_end = src_y_start + (y_end - y_start)
    src_x_start = max(0, -offset_x)
    src_x_end = src_x_start + (x_end - x_start)

    shifted[y_start:y_end, x_start:x_end] = mask[src_y_start:src_y_end, src_x_start:src_x_end]

    # Gaussian blur for softness
    kernel_size = blur * 2 + 1
    blurred = cv2.GaussianBlur(shifted.astype(np.float32), (kernel_size, kernel_size), blur / 3)

    # Scale by opacity
    shadow_alpha = (blurred * opacity).clip(0, 255).astype(np.uint8)

    # Build RGBA shadow (dark color)
    shadow = np.zeros((h, w, 4), dtype=np.uint8)
    shadow[:, :, 0] = 30   # R
    shadow[:, :, 1] = 35   # G
    shadow[:, :, 2] = 32   # B
    shadow[:, :, 3] = shadow_alpha

    return shadow

File: lib/renderer/shadow.ts
import { SHADOW_DEFAULTS } from "@/lib/constants";

export interface ShadowConfig {
  blur?: number;
  opacity?: number;
  offsetX?: number;
  offsetY?: number;
}

/**
 * Generate a soft drop shadow from the painting alpha mask.
 * Returns an RGBA buffer (same canvas dimensions) representing the shadow.
 */
export async function makeShadow(
  mask: Buffer,
  width: number,
  height: number,
  cfg: ShadowConfig = {},
): Promise<Buffer> {
  const sharp = (await import("sharp")).default;

  const blur = cfg.blur ?? SHADOW_DEFAULTS.blur;
  const opacity = cfg.opacity ?? SHADOW_DEFAULTS.opacity;
  const offsetX = cfg.offsetX ?? SHADOW_DEFAULTS.offsetX;
  const offsetY = cfg.offsetY ?? SHADOW_DEFAULTS.offsetY;

  // Shift mask by offset
  const shifted = Buffer.alloc(width * height);
  for (let y = 0; y < height; y++) {
    const srcY = y - offsetY;
    if (srcY < 0 || srcY >= height) continue;
    for (let x = 0; x < width; x++) {
      const srcX = x - offsetX;
      if (srcX < 0 || srcX >= width) continue;
      shifted[y * width + x] = mask[srcY * width + srcX];
    }
  }

  // Gaussian blur via sharp
  const sigma = Math.max(0.3, blur / 3);
  const blurred = await sharp(shifted, { raw: { width, height, channels: 1 } })
    .blur(sigma)
    .raw()
    .toBuffer();

  // Build RGBA shadow: dark color (30, 35, 32) with blurred alpha scaled by opacity
  const rgba = Buffer.alloc(width * height * 4);
  for (let i = 0; i < width * height; i++) {
    rgba[i * 4 + 0] = 30;
    rgba[i * 4 + 1] = 35;
    rgba[i * 4 + 2] = 32;
    rgba[i * 4 + 3] = Math.round(blurred[i] * opacity);
  }

  return rgba;
}
